pkg/cloudprovider/azure.go

```diff
@@ -1,33 +1,121 @@
+
+	tenantIDEnvVar       = "AZURE_TENANT_ID"
+	subscriptionIDEnvVar = "AZURE_SUBSCRIPTION_ID"
+	clientIDEnvVar       = "AZURE_CLIENT_ID"
+	clientSecretEnvVar   = "AZURE_CLIENT_SECRET"
+	cloudNameEnvVar      = "AZURE_CLOUD_NAME"
+)
+
+// Registry Credentials Secret
+const (
+	storageAccountKeyConfigKey = "storageAccountKey"
+)
+
 type AzureProvider struct {
 	BaseProvider
-	StorageAccount string
-	ResourceGroup  string
-	APITimeout     string
+	StorageAccount   string
+	StorageContainer string
+	ResourceGroup    string
+	APITimeout       string
 }
 
 func (p *AzureProvider) UpdateBSL(bsl *velero.BackupStorageLocation) {
 	bsl.Spec.Provider = Azure
+	bsl.Spec.StorageType = velero.StorageType{
+		ObjectStorage: &velero.ObjectStorageLocation{
+			Bucket: p.StorageContainer,
+			Prefix: "velero",
+		},
+	}
+
+	bsl.Spec.Config = map[string]string{
+		"resourceGroup":  p.ResourceGroup,
+		"storageAccount": p.StorageAccount,
+	}
 }
 
 func (p *AzureProvider) UpdateVSL(vsl *velero.VolumeSnapshotLocation) {
 	vsl.Spec.Provider = Azure
+
+	vsl.Spec.Config = map[string]string{
+		"resourceGroup": p.ResourceGroup,
+		"apiTimeout":    p.APITimeout,
+	}
 }
 
-func (p *AzureProvider) UpdateCloudSecret(ecret, cloudSecret *kapi.Secret) {
+func (p *AzureProvider) UpdateCloudSecret(secret, cloudSecret *kapi.Secret) {
+	cloudSecret.Data = map[string][]byte{
+		"cloud": secret.Data[AzureCredentials],
+	}
 }
 
-func (p *AzureProvider) UpdateRegistrySecret(secret, registrySecret *kapi.Secret) {
+func (p *AzureProvider) UpdateRegistrySecret(secret, registrySecret *kapi.Secret) error {
+	// 1. Parse AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_CLIENT_SECRET, AZURE_SUBSCRIPTION_ID
+	cloudCredsMap, err := godotenv.Unmarshal(string(secret.Data[AzureCredentials]))
+	if err != nil {
+		return err
+	}
+
+	// 2. Ask Azure API for Storage Account Key
+	storageAccountKey, _, err := p.getStorageAccountKey(cloudCredsMap)
+	if err != nil {
+		return err
+	}
+
+	// 3. Set the StorageAccountKey on the registry secret
+	registrySecret.Data = map[string][]byte{
+		storageAccountKeyConfigKey: []byte(storageAccountKey),
+	}
+	return nil
 }
 
-func (p *AzureProvider) UpdateRegistryDC(deploymentconfig *appsv1.DeploymentConfig, name, dirName string) {
+func (p *AzureProvider) UpdateRegistryDC(dc *appsv1.DeploymentConfig, name, dirName string) {
+	dc.Spec.Template.Spec.Containers[0].Env = []kapi.EnvVar{
+		{
+			Name:  "REGISTRY_STORAGE",
+			Value: "azure",
+		},
+		{
+			Name:  "REGISTRY_STORAGE_AZURE_CONTAINER",
+			Value: p.StorageContainer,
+		},
+		{
+			Name:  "REGISTRY_STORAGE_AZURE_ACCOUNTNAME",
+			Value: p.StorageAccount,
+		},
+		{
+			Name: "REGISTRY_STORAGE_AZURE_ACCOUNTKEY",
+			ValueFrom: &kapi.EnvVarSource{
+				SecretKeyRef: &kapi.SecretKeySelector{
+					LocalObjectReference: kapi.LocalObjectReference{Name: name},
+					Key:                  storageAccountKeyConfigKey,
+				},
+			},
+		},
+	}
 }
 
 func (p *AzureProvider) Validate(secret *kapi.Secret) []string {
```

Comment:
Let's validate the credentials secret here:
1. Better aligned with how invalid configuration is reported during _validation_.  See AWS.
2. Simplify code.  Don't need to check everywhere the secret is used.

https://github.com/fusor/mig-controller/blob/master/pkg/cloudprovider/aws.go#L162

Reply:
I'm looking at our AWS implementation, it seems like what you're asking for is:
 - Read the data from the secret
 - Make sure all required values within secret are filled (this is all we can do for Azure as far as I can tell)

I'm not sure what you mean with `Simplify code. Don't need to check everywhere the secret is used.`, can you elaborate?